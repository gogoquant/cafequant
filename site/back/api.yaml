openapi: 3.0.0
# Added by API Auto Mocking Plugin
servers:
# Added by API Auto Mocking Plugin
  - description: Matrix Cloud AppStack API
    url: "https://account.matrixcloud.io"
info:
  description: This is a simple API
  version: "1.1.0"
  title: Matrix Cloud AppStack API [非网关]
  contact:
    email: "hongwei.mei@changhong.com"
  license:
    name: Changhong SSC 2.0
    url: ""
tags:
  - name: templates
    description: "应用模版管理"
paths:
  /pets:
    get:
      summary: List all pets
      operationId: listPets
      tags:
        - pets
      parameters:
        - name: limit
          in: query
          description: How many items to return at one time (max 100)
          required: false
          type: integer
          format: int32
      responses:
        "200":
          description: An paged array of pets
          headers:
            x-next:
              type: string
              description: A link to the next page of responses
          schema:
            $ref: '#/definitions/Pets'
        default:
          description: unexpected error
          schema:
            $ref: '#/definitions/Error'
    post:
      summary: Create a pet
      operationId: createPets
      tags:
        - pets
      responses:
        "201":
          description: Null response
        default:
          description: unexpected error
          schema:
            $ref: '#/definitions/Error'
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      tags:
        - pets
      parameters:
        - name: petId
          in: path
          required: true
          description: The id of the pet to retrieve
          type: string
      responses:
        "200":
          description: Expected response to a valid request
          schema:
            $ref: '#/definitions/Pets'
        default:
          description: unexpected error
          schema:
            $ref: '#/definitions/Error'
  /users:
    post:
      summary: Create an user
      operationId: createUsers
      tags:
        - users
      requestBody:
        description: "add user"
        content:
          application/json:
            schema:
              $ref: "#/definitions/User"   
      responses:
        "201":
          description: Null response
        default:
          description: unexpected error
          schema:
            $ref: '#/definitions/Error'
definitions:
  User:
    type: object
    required:
      - email
      - name
      - passwd
    properties:
      email:
        type: string
      name:
        type: string
      passwd:
        type: string
  Users:
     type: array
     items:
      $ref: '#/definitions/User'    
  Pet:
    required:
      - id
      - name
    properties:
      id:
        type: integer
        format: int64
      name:
        type: string
      tag:
        type: string
  Pets:
    type: array
    items:
      $ref: '#/definitions/Pet'
  Error:
    required:
      - code
      - message
    properties:
      code:
        type: integer
        format: int32
      message:
        type: string