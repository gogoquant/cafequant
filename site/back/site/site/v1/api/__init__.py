# -*- coding: utf-8 -*-
from __future__ import absolute_import

from core import RequestHandler
from .. import UserInfo
from ..validators import request_validate, response_filter

class ApiHandler(RequestHandler):
    on_initialize_decorators = [response_filter, request_validate]

    def get_current_user(self):
        authorization = self.request.headers.get('Authorization', '')
        user_id = self.request.headers.get('user_id')

        return UserInfo(user_id, authorization, self.blueprint)