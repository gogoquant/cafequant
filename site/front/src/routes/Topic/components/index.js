export { default as TopicContent } from './TopicContent';
export { default as TopicReplies } from './TopicReplies';
export { default as TopicEditor } from './TopicEditor';
