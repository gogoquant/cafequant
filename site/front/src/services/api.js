// request api base url
 export const BASE_URL = 'http://127.0.0.1:8080';
// export const BASE_URL = '/proxy';

// cors support
export const CORS = true;

export const TopicsApi = {
  topic: '/topic/:id',
  topics: '/topics',
  topicsUpdate: '/topics/update',
};

export const UserApi = {
  user: '/user/:name',
  accessToken: '/accesstoken',
  register: '/api/user/register',
};

export const ReplyApi = {
  reply: '/topic/:topic_id/replies',
  upReply: '/reply/:reply_id/ups',
};

export const TopicCollectApi = {
  userCollect: '/topic_collect/:name',
  collectTopic: '/topic_collect/collect',
  cancelCollect: '/topic_collect/de_collect',
};
