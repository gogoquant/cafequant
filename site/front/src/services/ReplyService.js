import { post } from './instance';
import { ReplyApi } from './api';

/**
 * add reply to Topic
 * post: /topic/:topic_id/replies
 *
 * @param {string} token - user token
 * @param {string} topicId - topic ID
 * @param {string} content - reply content
 * @param {string} [replyId] - if reply to another reply then add this param
 * @returns {Promise<*>}
 */
export const addReply = (token, topicId, content, replyId) => {
  const params = {
    accesstoken: token,
    content,
  };
  if (replyId !== undefined) {
    params.reply_id = replyId;
  }

  return post(ReplyApi.reply.replace(/:topic_id/, topicId), params);
};

/**
 * up for Topic's reply
 * post: /reply/:reply_id/ups
 *
 * @param {string} token - user token
 * @param {string} replyId - reply to up
 * @returns {Promise<*>}
 */
export const upReply = (token, replyId) => {
  const params = {
    accesstoken: token,
  };
  return post(ReplyApi.upReply.replace(/:reply_id/, replyId), params);
};
