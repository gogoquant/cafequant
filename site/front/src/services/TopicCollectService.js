import { get, post } from './instance';
import { TopicCollectApi } from './api';

/**
 * get user's Topic collect
 * get: /topic_collect/:name
 *
 * @param {string} name - username
 * @returns {Promise<*>}
 */
export const getUserCollect = name => {
  return get(TopicCollectApi.userCollect.replace(/:name/, name));
};

/**
 * collect a Topic
 * post: /topic_collect/collect
 *
 * @param {string} token - user token
 * @param {string} id - topic ID
 * @returns {Promise<*>}
 */
export const collectTopic = (token, id) => {
  const params = {
    accesstoken: token,
    topic_id: id,
  };
  return post(TopicCollectApi.collectTopic, params);
};

/**
 * cancel collect a Topic
 * post: /topic_collect/de_collect
 *
 * @param {string} token - user token
 * @param {string} id - topic ID
 * @returns {Promise<*>}
 */
export const cancelTopic = (token, id) => {
  const params = {
    accesstoken: token,
    topic_id: id,
  };
  return post(TopicCollectApi.cancelCollect, params);
};
