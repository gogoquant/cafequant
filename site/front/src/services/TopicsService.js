import { userUtils } from 'utils';
import { get, post } from './instance';
import { TopicsApi } from './api';

/**
 * get Topics info
 * get: /topics
 *
 * @param {string} tab - topics tab
 * @param {number} page - topics page
 * @param {number} limit - topics page size
 * @returns {Promise<*>}
 */
export const getTopics = (tab, page, limit) => {
  const params = {
    tab,
    page,
    limit,
    mdrender: 'true',
  };
  return get(TopicsApi.topics, params);
};

/**
 * post Topics or update Topics
 * post: /topics for post a new Topics
 * post: /topics/update for update Topics
 *
 * @param {string} title - Topics title
 * @param {string} tab - Topics tab
 * @param {string} content - Topics content
 * @param {string} [id] - if update Topics then add this id param
 * @returns {Promise<*>}
 */
export const postTopics = (title, tab, content, id) => {
  const isPost = id === undefined;
  const { token: accesstoken } = userUtils.getUser();
  const params = {
    accesstoken,
    title,
    tab,
    content,
  };
  if (isPost) {
    return post(TopicsApi.topics, params);
  } else {
    params.topic_id = id;
    return post(TopicsApi.topicsUpdate, params);
  }
};

/**
 * get Topic info
 * get: /topic/:id
 *
 * @param {string} id - topic ID
 * @param {string} token - user token
 * @returns {Promise<*>}
 */
export const getTopic = (id, token) => {
  const params = { accesstoken: token, mdrender: 'false' };
  return get(TopicsApi.topic.replace(/:id/, id), params);
};
