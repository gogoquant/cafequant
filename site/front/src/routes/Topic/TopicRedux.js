import { Map, List, fromJS } from 'immutable';
import { createAction, handleActions } from 'redux-actions';
import { TopicsService, TopicCollectService, ReplyService } from 'services';

const LOAD_TOPIC = 'routes/Topic/LOAD_TOPIC';
const LOAD_TOPIC_SUCCESS = 'routes/Topic/LOAD_TOPIC_SUCCESS';
const LOAD_TOPIC_ERROR = 'routes/Topic/LOAD_TOPIC_ERROR';

const POST_COLLECT_TOPIC = 'routes/Topic/POST_COLLECT_TOPIC';
const POST_COLLECT_TOPIC_SUCCESS = 'routes/Topic/POST_COLLECT_TOPIC_SUCCESS';
const POST_COLLECT_TOPIC_ERROR = 'routes/Topic/POST_COLLECT_TOPIC_ERROR';

const UP_TOPIC_REPLY = 'routes/Topic/UP_TOPIC_REPLY';
const UP_TOPIC_REPLY_SUCCESS = 'routes/Topic/UP_TOPIC_REPLY_SUCCESS';
const UP_TOPIC_REPLY_ERROR = 'routes/Topic/UP_TOPIC_REPLY_ERROR';

const initialState = Map({
  loading: false,
  loadingCollect: false,
  loadingUp: false,

  error: false,
  collectError: false,
  upError: false,

  topicData: Map({}),
  repliesData: Map({
    entities: Map({}),
    result: List([]),
  }),
  isCollect: false,
});

const reducer = handleActions(
  {
    [LOAD_TOPIC]: state =>
      state
        .set('loading', true)
        .set('error', false)
        .set('topicData', Map({}))
        .set(
          'repliesData',
          Map({
            entities: Map({}),
            result: List([]),
          }),
        ),

    [LOAD_TOPIC_SUCCESS]: (
      state,
      { payload: { topicData, repliesData, isCollect } },
    ) =>
      state
        .set('loading', false)
        .set('topicData', topicData)
        .set('repliesData', repliesData)
        .set('isCollect', isCollect),

    [LOAD_TOPIC_ERROR]: state => state.set('loading', false).set('error', true),

    [POST_COLLECT_TOPIC]: state =>
      state.set('loadingCollect', true).set('collectError', false),

    [POST_COLLECT_TOPIC_SUCCESS]: state =>
      state
        .set('loadingCollect', false)
        .updateIn(['isCollect'], value => !value),

    [POST_COLLECT_TOPIC_ERROR]: state =>
      state.set('loadingCollect', false).set('collectError', true),

    [UP_TOPIC_REPLY]: state =>
      state.set('loadingUp', true).set('upError', false),

    [UP_TOPIC_REPLY_SUCCESS]: (state, { payload: { replyId, isUp } }) =>
      state
        .set('loadingUp', false)
        .updateIn(
          ['repliesData', 'entities', replyId, 'is_uped'],
          value => !value,
        )
        .updateIn(
          ['repliesData', 'entities', replyId, 'ups'],
          value => (isUp ? value + 1 : value - 1),
        ),

    [UP_TOPIC_REPLY_ERROR]: state =>
      state.set('loadingUp', false).set('upError', true),
  },
  initialState,
);

export const loadTopic = createAction(LOAD_TOPIC);
export const loadTopicSuccess = createAction(LOAD_TOPIC_SUCCESS);
export const loadTopicError = createAction(LOAD_TOPIC_ERROR);

export const postCollectTopic = createAction(POST_COLLECT_TOPIC);
export const postCollectTopicSuccess = createAction(POST_COLLECT_TOPIC_SUCCESS);
export const postCollectTopicError = createAction(POST_COLLECT_TOPIC_ERROR);

export const upTopicReply = createAction(UP_TOPIC_REPLY);
export const upTopicReplySuccess = createAction(UP_TOPIC_REPLY_SUCCESS);
export const upTopicReplyError = createAction(UP_TOPIC_REPLY_ERROR);

export const getTopicData = (id, token, callback) => async dispatch => {
  dispatch(loadTopic());
  try {
    const {
      data: { data },
    } = await TopicsService.getTopic(id, token);

    const { replies: repliesArray, ...topic } = data;
    const repliesLen = repliesArray.length;
    const repliesResult = [];
    const repliesEntities = {};
    repliesArray.forEach(item => {
      repliesResult.push(item.id);
      repliesEntities[item.id] = {
        ...item,
        ups: item.ups.length,
      };
    });
    const replies = {
      result: repliesResult,
      entities: repliesEntities,
    };
    const { is_collect: isCollect, ...topicRest } = topic;
    const topicData = fromJS(topicRest);
    const repliesData = fromJS(replies);
    dispatch(loadTopicSuccess({ topicData, repliesData, isCollect }));
    callback(repliesLen);
  } catch (e) {
    dispatch(loadTopicError());
  }
};

/**
 * collect topic when click collect button
 *
 * @param {bool} isCollect - true: collect topic, false: cancel collect topic
 * @param {string} token
 * @param {string} id
 * @returns {Function}
 */
export const collectTopic = (isCollect, token, id) => async dispatch => {
  dispatch(postCollectTopic());
  try {
    let data;
    if (isCollect) {
      data = await TopicCollectService.collectTopic(token, id);
    } else {
      data = await TopicCollectService.cancelTopic(token, id);
    }
    const {
      data: { success },
    } = data;
    if (success) {
      dispatch(postCollectTopicSuccess());
    } else {
      dispatch(postCollectTopicError());
    }
  } catch (e) {
    dispatch(postCollectTopicError());
  }
};

export const upReply = (token, replyId) => async dispatch => {
  dispatch(upTopicReply());
  try {
    const {
      data: { success, action },
    } = await ReplyService.upReply(token, replyId);
    if (success) {
      dispatch(upTopicReplySuccess({ replyId, isUp: action === 'up' }));
    } else {
      dispatch(upTopicReplyError());
    }
  } catch (e) {
    dispatch(upTopicReplyError());
  }
};

export default reducer;
