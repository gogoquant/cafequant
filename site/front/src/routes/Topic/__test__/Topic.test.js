import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import { instance } from 'services/instance';
import { ReplyApi } from 'services/api';
import toJson from 'enzyme-to-json';
import { TopicComponent } from '../Topic';

describe('<Topic />', () => {
  const topicProps = {
    topicTitle: 'title',
    currentUser: {
      id: '007',
    },
    error: false,
    getTopicData: jest.fn(),
  };
  const routeProps = {
    match: {
      params: {
        id: '123456',
      },
    },
  };

  it('should render correctly', () => {
    const tree = shallow(<TopicComponent {...topicProps} {...routeProps} />);
    expect(toJson(tree, { noKey: true })).toMatchSnapshot();
  });

  it('should render default title', () => {
    topicProps.topicTitle = undefined;
    const tree = shallow(<TopicComponent {...topicProps} {...routeProps} />);
    expect(toJson(tree, { noKey: true })).toMatchSnapshot();
  });

  it('should render <Exception />', () => {
    const props = {
      currentUser: {
        id: '007',
      },
      error: true,
      getTopicData: jest.fn(),
    };
    const tree = shallow(<TopicComponent {...props} {...routeProps} />);
    expect(toJson(tree, { noKey: true })).toMatchSnapshot();
  });

  describe('<TopicComponent />', () => {
    const iconPrefix = '.ant-notification-notice-icon';
    const props = {
      error: true,
      currentUser: {
        token: '123',
      },
      getTopicData: jest.fn(),
      upReply: jest.fn(),
    };

    let wrapper;
    let mockAxios;

    beforeEach(() => {
      wrapper = shallow(<TopicComponent {...props} {...routeProps} />);
      wrapper.setState({ contentValue: 'content' });
      mockAxios = sinon.stub(instance, 'post');
    });

    afterEach(() => {
      mockAxios.restore();
    });

    it('should change state when call onReplyPageChange()', () => {
      wrapper.instance().onReplyPageChange(2);
      expect(wrapper.state().replyPage).toEqual(2);
    });

    it('should change state when call onReplySizeChange()', () => {
      wrapper.instance().onReplySizeChange(2, 20);
      expect(wrapper.state().replyPage).toEqual(2);
      expect(wrapper.state().replySize).toEqual(20);
    });

    it('should call upReply() when click up', () => {
      wrapper.instance().onClickUp(2);
      expect(props.upReply).toHaveBeenCalledTimes(1);
    });

    it('should not call upReply() when token is undefined', () => {
      const ownProps = {
        ...props,
        currentUser: {
          token: undefined,
        },
        upReply: jest.fn(),
      };
      const ownWrapper = shallow(
        <TopicComponent {...ownProps} {...routeProps} />,
      );
      ownWrapper.instance().onClickUp(2);
      expect(ownProps.upReply).toHaveBeenCalledTimes(0);
    });

    it('should change state when call onEditorChange()', () => {
      wrapper.instance().onEditorChange('content123');
      expect(wrapper.state().contentValue).toEqual('content123');
    });

    it('should show error when contentValue length is 0', async () => {
      wrapper.setState({ contentValue: '' });
      await wrapper
        .instance()
        .onClickAddReply()
        .then(() => {
          expect(document.querySelectorAll(`${iconPrefix}-error`).length).toBe(
            1,
          );
        });
    });

    it('addReply() success', async () => {
      mockAxios
        .withArgs(
          ReplyApi.reply.replace(/:topic_id/, routeProps.match.params.id),
        )
        .resolves({ status: 200, data: { success: true, reply_id: '321' } });

      await wrapper
        .instance()
        .onClickAddReply()
        .then(() => {
          expect(wrapper.state().contentValue).toEqual('');
          expect(wrapper.state().replyId).toEqual('321');
          expect(wrapper.state().replyLoading).toBeFalsy();
        });
    });

    it('addReply() service return success: false', async () => {
      mockAxios
        .withArgs(
          ReplyApi.reply.replace(/:topic_id/, routeProps.match.params.id),
        )
        .resolves({ status: 200, data: { success: false } });

      await wrapper
        .instance()
        .onClickAddReply()
        .then(() => {
          expect(
            document.querySelectorAll(`${iconPrefix}-warning`).length,
          ).toBe(1);
          expect(wrapper.state().replyLoading).toBeFalsy();
        });
    });

    it('addReply() service error reject', async () => {
      mockAxios
        .withArgs(
          ReplyApi.reply.replace(/:topic_id/, routeProps.match.params.id),
        )
        .rejects();

      await wrapper
        .instance()
        .onClickAddReply()
        .then(() => {
          expect(
            document.querySelectorAll(`${iconPrefix}-warning`).length,
          ).toBe(2);
          expect(wrapper.state().replyLoading).toBeFalsy();
        });
    });

    it('addReply() success and jump to reply anchor', () => {
      const ownProps = {
        ...props,
        getTopicData: jest.fn((a, b, fn) => fn(5)),
      };
      const ownWrapper = shallow(
        <TopicComponent {...ownProps} {...routeProps} />,
      );
      ownWrapper.setState({ replyId: '110' });
      ownWrapper.instance().getTopicData();
      expect(ownWrapper.state().replyPage).toEqual(1);
    });

    it('addReply() success and change replyPage and jump to reply anchor', () => {
      const ownProps = {
        ...props,
        getTopicData: jest.fn((a, b, fn) => fn(11)),
      };
      const ownWrapper = shallow(
        <TopicComponent {...ownProps} {...routeProps} />,
      );
      ownWrapper.setState({ replyId: '110' });
      ownWrapper.instance().getTopicData();
      expect(ownWrapper.state().replyPage).toEqual(2);
    });
  });
});
