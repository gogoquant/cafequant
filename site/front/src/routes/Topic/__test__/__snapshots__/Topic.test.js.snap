// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Topic /> should render <Exception /> 1`] = `<Exception />`;

exports[`<Topic /> should render correctly 1`] = `
<Fragment>
  <Title
    title="title"
  />
  <withRouter(Connect(TopicContentComponent)) />
  <ConnectFunction
    current={1}
    onClickUp={[Function]}
    onReplyPageChange={[Function]}
    onReplySizeChange={[Function]}
    pageSize={10}
  />
  <TopicEditor
    loading={false}
    onClickReply={[Function]}
    onEditorChange={[Function]}
    value=""
  />
</Fragment>
`;

exports[`<Topic /> should render default title 1`] = `
<Fragment>
  <Title
    title="话题"
  />
  <withRouter(Connect(TopicContentComponent)) />
  <ConnectFunction
    current={1}
    onClickUp={[Function]}
    onReplyPageChange={[Function]}
    onReplySizeChange={[Function]}
    pageSize={10}
  />
  <TopicEditor
    loading={false}
    onClickReply={[Function]}
    onEditorChange={[Function]}
    value=""
  />
</Fragment>
`;
